contracts/contracts/pausing/Pauser.sol

```diff
@@ -0,0 +1,172 @@
+    bytes32 public constant TEMP_PAUSER_ROLE = keccak256("TEMP_PAUSER_ROLE");
+
+    /**
+     * @dev The current state of the contract.
+     */
+    PauseState public pauseState;
+
+    /**
+     * @dev The duration of a temp-pause in seconds.
+     */
+    uint256 public expiryDuration;
+
+    /**
+     * @dev The expiry timestamp of the current temp-pause.
+     * If unpaused or confirmed, the expiry timestamp should be zero (0).
+     */
+    uint256 public currentExpiry;
+
+    /**
+     * @dev The address of the managed contract.
+     */
+    address public pausable;
+
+    /**
+     * @dev Verifies that the caller is a whitelisted account or the governor or strategist.
+     */
+    modifier onlyWhitelistedOrGovernorOrStrategist() {
+        require(
+            hasRole(TEMP_PAUSER_ROLE, msg.sender) || _isGovernorOrStrategist(),
+            "Caller is not whitelisted and is not the Strategist or Governor"
+        );
+        _;
+    }
+
+    modifier onlyGovernorOrStrategist() {
+        require(
+            _isGovernorOrStrategist(),
+            "Caller is not the Strategist or Governor"
+        );
+        _;
+    }
+
+    function initialize(address _pausable, uint256 _expiryDuration)
+        external
+        onlyGovernor
+        initializer
+    {
+        setPausable(_pausable);
+        setExpiryDuration(_expiryDuration);
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    function addToWhitelist(address _account)
+        external
+        onlyGovernorOrStrategist
+    {
+        _grantRole(TEMP_PAUSER_ROLE, _account);
+    }
+
+    function removeFromWhitelist(address _account)
+        external
+        onlyGovernorOrStrategist
+    {
+        _revokeRole(TEMP_PAUSER_ROLE, _account);
+    }
+
+    /**
+     * @dev Pauses the managed contract indefinitely
+     */
+    function pause() external onlyWhitelistedOrGovernorOrStrategist {
+        // Cannot pause if already paused
+        require(pauseState != PauseState.PAUSED, "Contract is already paused");
+
+        if (_isGovernorOrStrategist()) {
+            pauseState = PauseState.PAUSED;
+            currentExpiry = 0;
+        } else {
+            pauseState = PauseState.TEMP_PAUSED;
+            currentExpiry = block.timestamp + expiryDuration;
+            _revokeRole(TEMP_PAUSER_ROLE, msg.sender);
+            emit TempPaused(msg.sender, currentExpiry);
+        }
+
+        // Execute the pause action on the contract
+        Pausable(pausable).pause();
+    }
+
+    /**
+     * @dev Unpauses the managed contract and resets expiry to 0
+     */
+    function unpause() external {
+        // Cannot unpause if already unpaused
+        require(
+            pauseState != PauseState.UNPAUSED,
+            "Contract is already unpaused"
+        );
+
+        // Normal users can unpause when the contract has expired
+        if (!_isGovernorOrStrategist()) {
+            require(
+                block.timestamp > currentExpiry,
```

Comment:
Let's be explicit here and also require that the contract is in TEMP_PAUSED. Right now I think anyone could unpause a full pause because the expiry was set to zero.

Reply:
True. Very important catch. I'll add tests to cover that case as well.